lhotse/augmentation/torchaudio.py

```diff
@@ -0,0 +1,111 @@
+    """
+    Class-style wrapper for torchaudio SoX effect chains.
+    It should be initialized with a config-like list of items that define SoX effect to be applied.
+    It supports sampling randomized values for effect parameters through the ``RandomValue`` wrapper.
+
+    Example:
+        >>> audio = np.random.rand(16000)
+        >>> augment_fn = SoxEffectTransform(effects=[
+        >>>    ['reverb', 50, 50, RandomValue(0, 100)],
+        >>>    ['speed', RandomValue(0.9, 1.1)],
+        >>>    ['rate', 16000],
+        >>> ])
+        >>> augmented = augment_fn(audio, 16000)
+
+    See SoX manual or ``torchaudio.sox_effects.effect_names()`` for the list of possible effects.
+    The parameters and the meaning of the values are explained in SoX manual/help.
+    """
+
+    def __init__(self, effects: EffectsList):
+        super().__init__()
+        self.effects = effects
+
+    def __call__(self, tensor: Union[torch.Tensor, np.ndarray], sampling_rate: int):
+        if isinstance(tensor, np.ndarray):
+            tensor = torch.from_numpy(tensor)
+        effects = self.sample_effects()
+        augmented, new_sampling_rate = torchaudio.sox_effects.apply_effects_tensor(tensor, sampling_rate, effects)
+        assert augmented.shape[0] == tensor.shape[0], "Lhotse does not support modifying the number " \
+                                                      "of channels during data augmentation."
+        assert sampling_rate == new_sampling_rate, \
+            f"Lhotse does not support changing the sampling rate during data augmentation. " \
+            f"The original SR was '{sampling_rate}', after augmentation it's '{new_sampling_rate}'."
+        # Matching shapes after augmentation -> early return.
+        if augmented.shape[1] == tensor.shape[1]:
+            return augmented
+        # We will truncate/zero-pad the signal if the number of samples has changed to mimic
+        # the WavAugment behavior that we relied upon so far.
+        resized = torch.zeros_like(tensor)
+        if augmented.shape[1] > tensor.shape[1]:
+            resized = augmented[:, :tensor.shape[1]]
+        else:
+            resized[:, :augmented.shape[1]] = augmented
+        return resized
+
+    def sample_effects(self) -> List[List[str]]:
+        """
+        Resolve a list of effects, replacing random distributions with samples from them.
+        It converts every number to string to match the expectations of torchaudio.
+        """
+        return [
+            [
+                str(item.sample() if isinstance(item, RandomValue) else item)
+                for item in effect
+            ]
+            for effect in self.effects
+        ]
+
+
+def speed(sampling_rate: int) -> List[List[str]]:
+    return [
+        # Random speed perturbation factor between 0.9x and 1.1x the original speed
+        ['speed', RandomValue(0.9, 1.1)],
+        ['rate', sampling_rate],  # Resample back to the original sampling rate (speed changes it)
```

Comment:
Looks like this line makes the running hang. It works without this line.

Reply:
😫

I'll have a look

Reply:
I was able to replicate the issue and add a unit test that causes it. I submitted the issue to torchaudio here: https://github.com/pytorch/audio/issues/1021

Reply:
BTW, if replacing RandomValue() above with a function  `_get_value(factor)` where `factor` is simply returned, the running hangs as well.

Reply:
Is the function defined as a closure (i.e. within another function) and captures some variable outside of its scope? That could explain it... Otherwise, I don't know.

Reply:
No, it's a independent function outside of any other functions.

https://github.com/freewym/espresso/pull/45/commits/e995f0d61ea9777a366a5a2d3b0bbe9eef900169#diff-3724bd15a663fd58b5fc92daa096bc9eb0c0a96a4c070a1a9e999be0682c985aR298

Reply:
@freewym some good news, if you create executor like: `ProcessPoolExecutor(..., mp_context=multiprocessing.get_context("spawn"))` it solves the segfault/hanging problem. Could you try? If it works I'll go on and merge this

Reply:
Credits to @mthrok for suggesting this

Reply:
(to make it clear: it works for me on the grid, on my mac, and in GitHub CI, so it should be okay)

Reply:
Yeah, it works!

Reply:
🎉